provider/core/src/refcount.rs

```diff
@@ -0,0 +1,144 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use core::ffi::c_void;
+use core::fmt::Debug;
+use core::mem::size_of;
+use core::{mem::ManuallyDrop, ptr::NonNull};
+use alloc::sync::Arc;
+use alloc::{string::String, rc::Rc};
+use alloc::boxed::Box;
+use stable_deref_trait::{StableDeref, CloneStableDeref};
+
+type WrappedType = Option<Box<[u8]>>;
+
+// Safe as a sentinel because 0x1 is not a valid usize address
+const SENTINEL: *const WrappedType = 1usize as *const WrappedType;
+
+/// A thin wrapper over [`NonNull`] that works with `*const T`
+#[derive(Debug)]
+#[repr(transparent)]
+pub(crate) struct NonNullConst<T>(NonNull<T>);
+
+impl<T> NonNullConst<T> {
+    #[inline]
+    pub const fn new_sentinel() -> Self {
+        // Safety: `SENTINEL` is non-null
+        Self(unsafe { NonNull::new_unchecked(SENTINEL as *mut T) })
+    }
+    #[inline]
+    pub fn new_from_rc(rc: Rc<T>) -> Self {
+        // Safety: `Rc::into_raw` returns a non-null pointer
+        Self(unsafe { NonNull::new_unchecked(Rc::into_raw(rc) as *mut T) })
+    }
+    #[inline]
+    pub fn new_from_arc(rc: Arc<T>) -> Self {
+        // Safety: `Arc::into_raw` returns a non-null pointer
+        Self(unsafe { NonNull::new_unchecked(Arc::into_raw(rc) as *mut T) })
+    }
+    #[inline]
+    fn into_raw(&mut self) -> Option<*const T> {
+        let ptr = self.0.as_ptr() as *const T;
+        if ptr == SENTINEL as *const T {
+            None
+        } else {
+            Some(ptr)
+        }
+    }
+    /// Safety: this `NonNullConst` must have been created with
+    /// `new_from_rc` or `new_sentinel`
+    #[inline]
+    pub unsafe fn into_assume_rc(&mut self) -> Option<Rc<T>> {
```

Comment:
this can be written in terms of `into_raw` at the call site

Reply:
It is now `take()` everywhere. Note that I cannot move from `NonNull` because it doesn't implement `Drop`. It is easier to write it in terms of `&mut self`.